### PR TITLE
feat(network): add directPeers support for GossipSub

### PR DESCRIPTION
**Pull request by catwith1hat**
Add support for GossipSub direct peers, allowing nodes to maintain permanent mesh connections without GRAFT/PRUNE negotiation. This enables proper peering agreements with other clients like Nimbus.

Direct peers can be configured via CLI:
  --directPeers /ip4/192.168.1.1/tcp/9000/p2p/16Uiu2HAm...

Both peers must configure each other as direct peers for the feature to work properly (reciprocal configuration).

**Motivation**

Direct peers are not supposed to send GRAFT/PRUNE messages. Pointing other CLs to Lodestar makes the other CLs receive these messages. Nimbus in response spams its logs with warning on each GRAFT/PRUNE message.

**Description**

The underlying p2p library already support direct peers. The PR is plumbing this feature to the a cmdline.

**AI Assistance Disclosure**

- [x] External Contributors: I have read the [contributor guidelines](https://github.com/ChainSafe/lodestar/blob/unstable/CONTRIBUTING.md#ai-assistance-notice) and disclosed my usage of AI below.

This code was generated by Claude as disclosed in the Co-Author line. Frankly, given that this is mostly plumbing there wasn't much to object to. More importantly, I pushed a modified binary to my production environment and verified that:
* Lodestar launches with the new cmdline
* Lodestar prints "Adding direct peer" for each peer given.
* Lodestar syncs and produces attestations.
* Nimbus stops complaining that Lodestar GRAFT/PRUNE messages.

### CODE REVIEW

**Review by lodekeeper — APPROVED** [review]
Nice clean implementation! This is mostly plumbing the existing libp2p direct peers feature to the CLI, which is the right approach.

**Review notes:**

✅ **parseDirectPeers function** - Good error handling with warnings for invalid multiaddrs. Proper extraction of peer ID and address from multiaddr.

✅ **Test coverage** - Comprehensive tests covering valid/invalid inputs, multiple peers, DNS multiaddrs, and edge cases.

✅ **CLI option** - The comma coerce pattern is consistent with existing code (e.g., `logOptions.ts`).

✅ **Documentation** - Clear description of the feature and reciprocal configuration requirement.

**Minor observation:**
The `logger.info` for each direct peer is fine for startup-time configuration logging. Just noting that most runtime peer logs use `debug` level, but startup config at `info` seems appropriate here.

Good to ship! 🚀

---
🤖 Review assisted by AI

File: packages/cli/src/options/beaconNodeOptions/network.ts [hunk 1]
```diff
@@ -259,6 +261,17 @@ export const options: CliCommandOptions<NetworkArgs> = {
...
 
+  directPeers: {
+    type: "array",
+    description:
+      "Direct peers for GossipSub mesh. These peers maintain permanent connections without GRAFT/PRUNE. " +
+      "Format: multiaddr with peer ID, e.g., /ip4/192.168.1.1/tcp/9000/p2p/16Uiu2HAmKLhW7... " +
+      "Both peers must configure each other as direct peers for the feature to work properly.",
+    defaultDescription: "[]",
+    group: "network",
+    coerce: (args: string[]) => args.flatMap((item) => item.split(",")),
```
Comment (lodekeeper) [inline packages/cli/src/options/beaconNodeOptions/network.ts]: **Suggestion:** The coerce function may receive `undefined` in some yargs edge cases, and comma-split can produce whitespace or empty strings:

```suggestion
    coerce: (args: string[] | undefined) =>
      (args ?? []).flatMap((item) => item.split(",")).map((s) => s.trim()).filter(Boolean),
```

This handles:
- `undefined` input (yargs edge case)
- Whitespace around commas: `"a, b"` → `["a", "b"]`
- Trailing commas: `"a,b,"` → `["a", "b"]` (no empty string)

File: packages/beacon-node/src/network/gossip/gossipsub.ts [hunk 2]
```diff
@@ -381,3 +392,43 @@ function getForkBoundaryLabel(boundary: ForkBoundary): ForkBoundaryLabel {
...
+        logger.warn("Direct peer multiaddr must contain /p2p/ component with peer ID", {multiaddr: peerStr});
+        continue;
+      }
+
+      const peerId = peerIdFromString(peerIdStr);
+
+      // Get the address without the /p2p/ component
+      const addr = ma.decapsulate("/p2p/" + peerIdStr);
+
+      directPeers.push({
```
Comment (lodekeeper) [inline packages/beacon-node/src/network/gossip/gossipsub.ts]: **Suggestion:** The deprecated `/ipfs/` protocol alias (still seen in the wild) will cause `decapsulate("/p2p/...")` to fail since the multiaddr uses `/ipfs/` not `/p2p/`. Consider using protocol code instead:

```suggestion
      const addr = ma.decapsulateCode(421); // 421 = p2p protocol code, handles both /p2p/ and /ipfs/
```

Alternatively, explicitly reject `/ipfs/` with a clear error message.

**Review by lodekeeper — COMMENTED** [review]
Great implementation! A few minor suggestions to improve robustness:

**Comment by lodekeeper** [comment]
**CI Failure: Lint**

The lint check is failing due to formatting in `packages/cli/test/unit/options/beaconNodeOptions.test.ts`. The `directPeers` arrays should be single-line:

```ts
// Current (multiline - fails lint):
directPeers: [
  "/ip4/192.168.1.1/tcp/9000/p2p/16Uiu2HAkuWPWqF4W3aw9oo5Yw79v5muzBaaGTGKMmuqjPfEyfkwu",
],

// Expected (single-line):
directPeers: ["/ip4/192.168.1.1/tcp/9000/p2p/16Uiu2HAkuWPWqF4W3aw9oo5Yw79v5muzBaaGTGKMmuqjPfEyfkwu"],
```

You can fix this by running `pnpm lint --fix` locally.

---

The "Docs spellcheck" failure appears unrelated to this PR (no docs were changed) - likely a pre-existing issue or flaky test.

File: packages/beacon-node/src/network/gossip/gossipsub.ts (hunk 2 above)
Comment (catwith1hat) [reply packages/beacon-node/src/network/gossip/gossipsub.ts]: Addressed.

File: packages/cli/src/options/beaconNodeOptions/network.ts (hunk 1 above)
Comment (catwith1hat) [reply packages/cli/src/options/beaconNodeOptions/network.ts]: Addressed.

File: packages/beacon-node/src/network/gossip/gossipsub.ts [hunk 3]
```diff
@@ -381,3 +392,43 @@ function getForkBoundaryLabel(boundary: ForkBoundary): ForkBoundaryLabel {
...
+        id: peerId,
+        addrs: [addr],
+      });
+
+      logger.info("Added direct peer", {peerId: peerIdStr, addr: addr.toString()});
+    } catch (e) {
+      logger.warn("Failed to parse direct peer multiaddr", {multiaddr: peerStr, error: (e as Error).message});
+    }
+  }
```
[the remark this answers is not in the document]
Comment (catwith1hat) [reply packages/beacon-node/src/network/gossip/gossipsub.ts:465]: Addressed.

File: packages/cli/src/options/beaconNodeOptions/network.ts [hunk 4]
```diff
@@ -259,6 +261,21 @@ export const options: CliCommandOptions<NetworkArgs> = {
     group: "network",
   },
 
+  directPeers: {
+    type: "array",
+    description:
+      "Direct peers for GossipSub mesh. These peers maintain permanent connections without GRAFT/PRUNE. " +
+      "Format: multiaddr with peer ID, e.g., /ip4/192.168.1.1/tcp/9000/p2p/16Uiu2HAmKLhW7... " +
```
Comment (nflaig) [inline packages/cli/src/options/beaconNodeOptions/network.ts]: this value needs to be surrounded by backticks, otherwise it will be interpreted as text which fails the docs check right now
```suggestion
      "Format: multiaddr with peer ID, e.g., `/ip4/192.168.1.1/tcp/9000/p2p/16Uiu2HAmKLhW7...` " +
```

File: packages/beacon-node/src/network/gossip/gossipsub.ts [hunk 5]
```diff
@@ -381,3 +392,50 @@ function getForkBoundaryLabel(boundary: ForkBoundary): ForkBoundaryLabel {
...
+      if (!peerIdStr) {
+        logger.warn("Direct peer multiaddr must contain /p2p/ component with peer ID", {multiaddr: peerStr});
+        continue;
+      }
+
+      try {
+        const peerId = peerIdFromString(peerIdStr);
+
+        // Get the address without the /p2p/ component (use protocol code to handle /ipfs/ alias)
+        const addr = ma.decapsulateCode(421);
```
Comment (nflaig) [inline packages/beacon-node/src/network/gossip/gossipsub.ts]: why do we need to handle this case?

File: packages/beacon-node/src/network/gossip/gossipsub.ts [hunk 6]
```diff
@@ -381,3 +392,50 @@ function getForkBoundaryLabel(boundary: ForkBoundary): ForkBoundaryLabel {
...
 }
+
+/**
+ * Parse direct peer multiaddr strings into AddrInfo objects for GossipSub.
+ * Direct peers maintain permanent mesh connections without GRAFT/PRUNE negotiation.
+ *
+ * Expected format: /ip4/192.168.1.1/tcp/9000/p2p/16Uiu2HAmKLhW7...
+ * The multiaddr must contain a /p2p/ component with the peer ID.
+ */
+export function parseDirectPeers(directPeerStrs: string[], logger: Logger): AddrInfo[] {
```
Comment (nflaig) [inline packages/beacon-node/src/network/gossip/gossipsub.ts:409]: would also be great if we can also support passing ENRs here and extract the multiaddr from it

**Comment by lodekeeper** [comment]
Responding to @nflaig's review:

**Re: backticks for multiaddr example**
Good catch! The backticks will prevent the docs parser from interpreting `/ip4/...` as regular text.

**Re: handling `/ipfs/` case**
The `/ipfs/` protocol alias was used before `/p2p/` became the standard. Some older node configs or documentation may still reference it. Without handling this, `decapsulate("/p2p/...")` would fail silently for those multiaddrs. Using the protocol code (421) handles both variants automatically.

That said, if you prefer to explicitly reject `/ipfs/` with a clear error message instead of silently accepting it, that's also reasonable — it would encourage operators to update their configs.

**Re: ENR support**
Great idea! ENRs contain the multiaddr + peer ID info we need. Could be a follow-up PR to keep this one focused, or @catwith1hat can add it here if they prefer. The implementation would involve detecting `enr:` prefix and using `@chainsafe/enr` to decode.

File: packages/beacon-node/src/network/gossip/gossipsub.ts (hunk 5 above)
Comment (nflaig) [reply packages/beacon-node/src/network/gossip/gossipsub.ts]: ah this was suggested here https://github.com/ChainSafe/lodestar/pull/8831#discussion_r2751707743, I don't think it's useful to handle this case

Comment (catwith1hat) [reply packages/beacon-node/src/network/gossip/gossipsub.ts]: Should I revert to only handle /p2p/?

Comment (nflaig) [reply packages/beacon-node/src/network/gossip/gossipsub.ts]: yeah that should be fine, I don't see why we need to handle `/ipfs/`

Comment (catwith1hat) [reply packages/beacon-node/src/network/gossip/gossipsub.ts]: Reverted to only handle p2p.

File: packages/beacon-node/src/network/gossip/gossipsub.ts [hunk 7]
```diff
@@ -381,3 +394,79 @@ function getForkBoundaryLabel(boundary: ForkBoundary): ForkBoundaryLabel {
...
+        }
+
+        directPeers.push({
+          id: peerId,
+          addrs: [multiaddrTCP],
+        });
+
+        logger.info("Added direct peer from ENR", {peerId: peerId.toString(), addr: multiaddrTCP.toString()});
+      } catch (e) {
+        logger.warn("Failed to parse direct peer ENR", {enr: peerStr, error: (e as Error).message});
```
Comment (nflaig) [inline packages/beacon-node/src/network/gossip/gossipsub.ts]: error should always be passed as 3rd arg


```suggestion
        logger.warn("Failed to parse direct peer ENR", {enr: peerStr}, e as Error);
```

File: packages/beacon-node/src/network/gossip/gossipsub.ts [hunk 8]
```diff
@@ -381,3 +394,79 @@ function getForkBoundaryLabel(boundary: ForkBoundary): ForkBoundaryLabel {
...
+          logger.info("Added direct peer", {peerId: peerIdStr, addr: addr.toString()});
+        } catch (e) {
+          logger.warn("Invalid peer ID in direct peer multiaddr", {
+            multiaddr: peerStr,
+            peerId: peerIdStr,
+            error: (e as Error).message,
+          });
+        }
+      } catch (e) {
+        logger.warn("Failed to parse direct peer multiaddr", {multiaddr: peerStr, error: (e as Error).message});
```
Comment (nflaig) [inline packages/beacon-node/src/network/gossip/gossipsub.ts]:
```suggestion
        logger.warn("Failed to parse direct peer multiaddr", {multiaddr: peerStr}, e as Error);
```

File: packages/beacon-node/src/network/gossip/gossipsub.ts [hunk 9]
```diff
@@ -381,3 +394,79 @@ function getForkBoundaryLabel(boundary: ForkBoundary): ForkBoundaryLabel {
...
+            id: peerId,
+            addrs: [addr],
+          });
+
+          logger.info("Added direct peer", {peerId: peerIdStr, addr: addr.toString()});
+        } catch (e) {
+          logger.warn("Invalid peer ID in direct peer multiaddr", {
+            multiaddr: peerStr,
+            peerId: peerIdStr,
+            error: (e as Error).message,
```
Comment (nflaig) [inline packages/beacon-node/src/network/gossip/gossipsub.ts]: move to 3rd arg

File: packages/beacon-node/test/unit/network/gossip/directPeers.test.ts [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+import {beforeEach, describe, expect, it, vi} from "vitest";
+import {SignableENR} from "@chainsafe/enr";
+import {Logger} from "@lodestar/utils";
+import {parseDirectPeers} from "../../../../src/network/gossip/gossipsub.js";
+
+describe("network / gossip / directPeers", () => {
+  let logger: Logger;
+
+  beforeEach(() => {
+    logger = {
```
Comment (nflaig) [inline packages/beacon-node/test/unit/network/gossip/directPeers.test.ts]: there is a `getMockedLogger` here https://github.com/ChainSafe/lodestar/blob/2d390f92ac80c928b3fbc89cd98f9af6495b47ee/packages/beacon-node/test/mocks/loggerMock.ts#L6

**Review by nflaig — COMMENTED** [review]
generally lgtm, just a few minor comments, leaving up final approval to @wemeetagain

**Review by nflaig — DISMISSED** [review]

**Comment by catwith1hat** [comment]
Just pushed the last revision. I hope that I have addressed all your comments.

**Comment by nflaig** [comment]
> Just pushed the last revision. I hope that I have addressed all your comments.

bad timing, I just pushed few changes before that to make the PR ready to merge, seems like my changes are force pushed out though

**Comment by nflaig** [comment]
btw as a general note, it's hard to follow as a reviewer if you force push, simpler to just push new commits

**Comment by nflaig** [comment]
> Just pushed the last revision. I hope that I have addressed all your comments.

thanks looks good, just pushed changes to fix lint, running CI now but should pass

**Review by nflaig — APPROVED** [review]
Looks great, thanks @catwith1hat, we've needed this for a while, would have been useful to have during the holesky rescue as well.

After we have this merged it should be pretty easy to resolve https://github.com/ChainSafe/lodestar/issues/7559. Might be useful to add direct/trusted peers without having to reboot the node, especially during non-finality.

Will see if I can get the api implemented for 1.40 release but will have your PR included for sure.

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.40.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.40.0) :tada:
